src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectTypes/Server/ServerData.cs

```diff
@@ -0,0 +1,805 @@
+                serverConfig.ConfigValue = this.currentState.MaxMemory;
+            }
+
+            if (changesMade)
+            {
+                server.Configuration.Alter(true);
+            }
+
+        }
+        #endregion
+
+        public void ApplyInfoToPrototype(ServerInfo serverInfo)
+        {
+            this.Name = serverInfo.Name;
+            this.Language = serverInfo.Language;
+            this.Memory = serverInfo.MemoryInMB;
+            this.OperatingSystem = serverInfo.OperatingSystem;
+            this.Platform = serverInfo.Platform;
+            this.Version = serverInfo.Version;
+            this.Processors = serverInfo.Processors;
+            this.Version = serverInfo.Version;
+            this.MaxServerMemory = serverInfo.MaxServerMemory;
+            this.MinServerMemory = serverInfo.MinServerMemory;
+        }
+
+        /// <summary>
+        /// Private class encapsulating the data that is changed by the UI.
+        /// </summary>
+        /// <remarks>
+        /// Isolating this data allows for an easy implementation of Reset() and
+        /// simplifies difference detection when committing changes to the server.
+        /// </remarks>
+        private class ServerPrototypeData : ICloneable
+        {
+            #region data members
+            private string serverName = string.Empty;
+            private string hardwareGeneration = String.Empty;
+            private string language = String.Empty;
+            private int memoryInMB = 0;
+            private string operatingSystem = String.Empty;
+            private string platform = String.Empty;
+            private int processors = 0;
+            private bool isClustered = false;
+            private bool isHadrEnabled = false;
+            private bool isPolyBaseInstalled = false;
+            private bool isXTPSupported = false;
+            private string product = String.Empty;
+            private string rootDirectory = String.Empty;
+            private string serverCollation = String.Empty;
+            private string version = String.Empty;
+            private int minMemory = 0;
+            private int maxMemory = 0;
+
+            private bool initialized = false;
+            private Server server = null;
+            private CDataContainer context = null;
+            private ServerConfigService configService = null;
+            private bool isYukonOrLater = false;
+            #endregion
+
+            #region Properties
+
+            // General properties
+
+
+            public string ServerName
+            {
+                get
+                {
+                    if (!this.initialized)
+                    {
+                        LoadData();
+                    }
+
+                    return this.serverName;
+                }
+
+                set
+                {
+                    System.Diagnostics.Debug.Assert(this.initialized, "unexpected property set before initialization");
```

Comment:
Remove all instances of Debug.Assert. We should either be throwing an exception or logging it and providing some appropriate default behavior.